substrate/frame/assets/src/functions.rs

```diff
@@ -228,20 +237,21 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		let account = Account::<T, I>::get(&id, who).ok_or(Error::<T, I>::NoAccount)?;
 		ensure!(!account.status.is_frozen(), Error::<T, I>::Frozen);
 
-		let amount = if let Some(frozen) = T::Freezer::frozen_balance(id, who) {
-			// Frozen balance: account CANNOT be deleted
-			let required =
-				frozen.checked_add(&details.min_balance).ok_or(ArithmeticError::Overflow)?;
-			account.balance.saturating_sub(required)
-		} else {
-			if keep_alive {
-				// We want to keep the account around.
-				account.balance.saturating_sub(details.min_balance)
-			} else {
-				// Don't care if the account dies
-				account.balance
-			}
+		let untouchable = match (
+			T::Holder::held_balance(id.clone(), who),
+			T::Freezer::frozen_balance(id.clone(), who),
+			keep_alive,
+		) {
+			(None, None, true) => details.min_balance,
+			(None, None, false) => Zero::zero(),
+			(maybe_held, maybe_frozen, _) => {
+				let held = maybe_held.unwrap_or_default();
+				let frozen = maybe_frozen.unwrap_or_default();
+				frozen.saturating_sub(held).max(details.min_balance)
+			},
 		};
```

Comment:
Related to [this comment](https://github.com/paritytech/polkadot-sdk/pull/4530/files/c5b103a2601923a42c34d76a07e9b78e4772efd7#r1699500389) @gui1117:

Maybe a reasonable addition here would be supporting forcibly killing an account when there's no `held` balances.

```suggestion
		let untouchable = match (
			T::Holder::held_balance(id.clone(), who),
			T::Freezer::frozen_balance(id.clone(), who),
			keep_alive,
		) {
			(None, maybe_frozen, keep_alive) => if keep_alive {
				maybe_frozen.unwrap_or_default().max(details.min_balance)
			} else {
				Zero::zero()
			},
			(Some(held), maybe_frozen, _) => 
				maybe_frozen.unwrap_or_default().saturating_sub(held).max(details.min_balance),
			},
		};
```

In this case, if `held` amount exists, the account can **never** be killed. However, if there's no `held` amount, it might be dropped below `untouchable`, therefore deleting the account.

Reply:
Update about this suggestion @gui1117: changing this implies changing breaking several unit tests that hold the previous assumptions, especially the following ones:

- `freezer_should_work`, with this error:
  ```rs
  ---- tests::freezer_should_work stdout ----
  thread 'tests::freezer_should_work' panicked at substrate/frame/assets/src/functions.rs:295:17:
  passed from reducible_balance; qed
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  ```
- `transfer_all_works_3`, with this error:
  ```rs
  ---- tests::transfer_all_works_3 stdout ----
  thread 'tests::transfer_all_works_3' panicked at substrate/frame/assets/src/tests.rs:839:9:
  Expected Ok(_). Got Err(
      Token(
          FundsUnavailable,
      ),
  )
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  ```

In that case, I'd prefer a second opinion (maybe @muharem or @joepetrowski can take a look at this change, and if it would be hugely impactful?)

Reply:
If it breaks tests it's obviously not acceptable.